src/visit/filter.rs

```diff
@@ -409,6 +425,40 @@ impl<'a, G, I, F> Iterator for EdgeFilteredEdges<'a, G, I, F>
     }
 }
 
+/// A filtered neighbors-directed iterator.
+pub struct EdgeFilteredNeighborsDirected<'a, G, F: 'a>
+where
+    G: IntoEdgesDirected,
+{
+    iter: G::EdgesDirected,
+    f: &'a F,
+    dir: Direction,
+}
+
+impl<'a, G, F> Iterator for EdgeFilteredNeighborsDirected<'a, G, F>
+where
+    F: FilterEdge<G::EdgeRef>,
+    G: IntoEdgesDirected,
+{
+    type Item = G::NodeId;
+    fn next(&mut self) -> Option<Self::Item> {
+        let f = self.f;
+        let dir = self.dir;
+        (&mut self.iter)
+            .filter_map(move |edge| {
+                if f.include_edge(edge) {
+                    match dir {
+                        Direction::Incoming => Some(edge.source()),
+                        Direction::Outgoing => Some(edge.target()),
```

Comment:
Is this producing the right neighbors for undirected graphs? I'm not sure it does. We need to follow this:

To follow the doc: https://docs.rs/petgraph/0.4.13/petgraph/visit/trait.IntoNeighborsDirected.html

The underlying IntoEdgesDirected, for undirected graphs, should already produce edges where the source is the starting node (`n` in the `neighbors_directed` new method in this PR) and the target is the other edge node.

Reply:
Thanks. I've got it working for undirected graphs without having to work with the IntoNeighborsDirected trait. What I do is persist the starting node `n`, and then look at both the source and target of each edge; the node that `!= n` is the neighbor, except when it's an edge to itself.